### PR TITLE
Pytorch style maxpool2d

### PR DESCRIPTION
**Pull request by Flowingsun007**
实现flow.MaxPoolNd(MaxPool1d，MaxPool2d、MaxPool3d)，对标pytorch：[torch.nn.MaxPoolNd](https://pytorch.org/docs/master/generated/torch.nn.MaxPool2d.html#maxpool2d)
包括：
- flow.nn.MaxPoolNd的前向、反向实现、测试用例等

### Op
 - [x] Op SetBatchAxisInferFn
 - [x] Op SetGetSbpFn
 - [x] Op SetInputArgModifyFn
 - [x] Op 反向梯度注册

### Kernel
 - [x] CPU in:float32 
 - [x] CPU in:float64
 - [x] CPU in:int32
 - [x] CPU in:int64
 - [ ] CPU in:int8

 - [x] GPU in:float32 
 - [x] GPU in:float64
 - [x] GPU in:int32
 - [x] GPU in:int64
 - [ ] GPU in:float16
 - [ ] GPU in:int8

### Python Wrapper
 - [x] Python API 参数检查及异常提示
 - [x] 接口注释
 - [x] Example 

### 测试
 - [x] 单机单卡  CPU Test Case 
 - [x] 单机单卡  GPU Test Case 
 - [x] 单机多卡  CPU Test Case 
 - [x] 单机多卡  GPU Test Case
 - [ ] 分布式  CPU Test Case  
 - [ ] 分布式  GPU Test Case  

## GPU 有效带宽
请参考 https://github.com/Oneflow-Inc/OneTeam/issues/167
理论带宽：
```shell
 Device to Device Bandwidth, 1 Device(s)
 PINNED Memory Transfers
   Transfer Size (Bytes)	Bandwidth(MB/s)
   33554432			513802.2 (501GB/s)
```

实际带宽：
```shell
I0104 09:18:41.817276 42957 kernel.cpp:89] PROFILER::KERNEL::CUDA_MEMORY_BANDWIDTH op_name: MaxPool2d_2 elapsed(ms): 0.351264 memory_size(Byte): 75301056 bandwidth(GB/s): 199.649
I0104 09:18:41.817589 42957 kernel.cpp:89] PROFILER::KERNEL::CUDA_MEMORY_BANDWIDTH op_name: MaxPool2d_2_grad elapsed(ms): 0.264832 memory_size(Byte): 125534496 bandwidth(GB/s): 441.461
```

## PR Checklist
 - [x] PR 标题语句通畅，明确表达 PR 内容，适合直接作为新版本发布时的 changelog
 - [x] 代码格式化 
 - [x] 已经本地编译通过
 - [x] 已本地针对改动测试
 - [x] 已添加 type 标签:(填写 type 标签名，如 `bug, enhancement, purge, feature, documentation`)
 - [x] 已添加 component 标签:(填写 component 标签名，如 `op, system, eager, build, xla, python, ci, test, tooling, onnx`)
 - [x] Draft 转正式 PR 前已请人 Review

### CODE REVIEW

**Comment by doombeaker** [comment]
OP PR 的描述，现在已经有正式的模板了，在这里：https://github.com/Oneflow-Inc/oneflow/blob/master/.github/PULL_REQUEST_TEMPLATE/op_template.md

File: oneflow/user/kernels/pooling_kernel.cu [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+            index_helper, elem_num, src, dest, indice_ptr, n_batch, n_channel, src_height,
+            src_width, dst_height, dst_width, return_indices, ceil_mode);
+  }
+};
+
+OF_PP_SEQ_PRODUCT_FOR_EACH_TUPLE(INSTANTIATE_POOLING_KERNEL_UTIL, (DeviceType::kGPU),
+                                 POOLING_DATA_TYPE_GPU_SEQ);
+
+}  // namespace oneflow
+#endif  // WITH_CUDA
```
Comment (leaves-zwx) [inline oneflow/user/kernels/pooling_kernel.cu:102]: newline

File: oneflow/user/kernels/pooling_kernel_util.cpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+Shape PoolingParams3D::GetXShape5D() const {
+  return Shape({batch_num_, channel_num_, x_3d_.at(0), x_3d_.at(1), x_3d_.at(2)});
+}
+
+Shape PoolingParams3D::GetYShape5D() const {
+  return Shape({batch_num_, channel_num_, y_3d_.at(0), y_3d_.at(1), y_3d_.at(2)});
+}
+
+}  // namespace oneflow
```
Comment (leaves-zwx) [inline oneflow/user/kernels/pooling_kernel_util.cpp:115]: newline

File: oneflow/user/kernels/pooling_kernel_util.h [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    }
+  }
+}
+
+#define INSTANTIATE_POOLING_KERNEL_UTIL(device_type_v, dtype_pair) \
+  template struct PoolingKernelUtil<device_type_v, OF_PP_PAIR_FIRST(dtype_pair)>;
+
+}  // namespace oneflow
+
+#endif  // ONEFLOW_USER_KERNELS_POOLING_KERNEL_UTIL_H_
```
Comment (leaves-zwx) [inline oneflow/user/kernels/pooling_kernel_util.h:206]: newline

**Review by leaves-zwx — COMMENTED** [review]
- 代码实现还是太复杂，比较混乱
- cuda 实现应该可以调用 cudnn 的 Pooling 接口
- 1d 和 3d 版本应该也可以顺带实现（不需要增加太多代码）
- 注意一下代码规范

**Comment by leaves-zwx** [comment]
pool op 比较适合 cpu kernel 和 gpu kernel 分开到两个源文件实现，*_kernel.cpp, *_kernel.cu

**Comment by Flowingsun007** [comment]
> 代码实现还是太复杂，比较混乱

这块，具体一点是哪里复杂不清楚啊？（代码逻辑参考了原有maxpooling的实现+pytorch的实现，感觉还是比较易懂的哈）

> * cuda 实现应该可以调用 cudnn 的 Pooling 接口

我看cudnn的Pooling接口的，Descriptor不支持dilation参数，应该是不好支持的吧？

> * 1d 和 3d 版本应该也可以顺带实现（不需要增加太多代码）

好的，1d和3d的我加一下

**Comment by Flowingsun007** [comment]
### maxpool op性能对比
oneflow-15，环境完全对齐下编译、执行：
```shell
# 编译
cmake .. -DTHIRD_PARTY_MIRROR=aliyun  -DCMAKE_BUILD_TYPE=Release  -DCMAKE_EXPORT_COMPILE_COMMANDS=1 -DBUILD_PROFILER=ON
make -j 48

# 设置环境变量
export ONEFLOW_PROFILER_KERNEL_PROFILE_CUDA_MEMORY_BANDWIDTH="yes"

# 执行测试案例(gpu+float)
# （参数保持一致：x.shape=(1, 6, 1024, 1024);kernel_size=(3,3);stride=1;padding=0;）
python /home/luyang/Oneflow/oneflow/oneflow/python/test/ops/test_maxpool2d.py
```
### 结果
flow.nn.max_pool2d：
```shell
# y= flow.nn.max_pool2d(
                #     input=x,
                #     ksize=kernel_size,
                #     strides=1,
                #     padding=((0,0), (0,0), (0,0), (0,0)),
                #     ceil_mode=False,
                #     data_format="NCHW",
                # )
I0123 09:48:21.081825  6028 kernel.cpp:89] PROFILER::KERNEL::CUDA_MEMORY_BANDWIDTH op_name: MaxPool2D_2 elapsed(ms): 2.1671 memory_size(Byte): 50233440 bandwidth(GB/s): 21.588
I0123 09:48:21.082507  6028 kernel.cpp:89] PROFILER::KERNEL::CUDA_MEMORY_BANDWIDTH op_name: MaxPool2D_2_grad elapsed(ms): 0.633888 memory_size(Byte): 100466880 bandwidth(GB/s): 147.608
```
flow.nn.MaxPool2d
```shell
# y = flow.nn.MaxPool2d(
                #     input=x,
                #     kernel_size=kernel_size,
                #     stride=stride,
                #     padding=padding,
                #     dilation=1,
                #     return_indices=False,
                #     ceil_mode=False,
                #     data_format="NCHW",
                # )
I0123 09:51:50.247990 12399 kernel.cpp:89] PROFILER::KERNEL::CUDA_MEMORY_BANDWIDTH op_name: MaxPool2d_2 elapsed(ms): 0.737024 memory_size(Byte): 100368672 bandwidth(GB/s): 126.828
I0123 09:51:50.248317 12399 kernel.cpp:89] PROFILER::KERNEL::CUDA_MEMORY_BANDWIDTH op_name: MaxPool2d_2_grad elapsed(ms): 0.283744 memory_size(Byte): 150602112 bandwidth(GB/s): 494.316
```

**Comment by leaves-zwx** [comment]
上面的性能测试没看懂是谁跟谁对比，写的是 flow.nn.max_pool2d vs. flow.nn.MaxPool2d ?

**Comment by Flowingsun007** [comment]
> 上面的性能测试没看懂是谁跟谁对比，写的是flow.nn.max_pool2d与flow.nn.MaxPool2d？

就是新开发的torch风格的flow.nn.MaxPool2d和之前oneflow已经存在的flow.nn.max_pool2d对比

File: oneflow/user/kernels/pooling_kernel_util.h [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+
+template<typename T>
+OF_DEVICE_FUNC void Maxpool2dFarwardCompute(
+    const NdIndexOffsetHelper<int64_t, 4> index_helper, int64_t elem_num, T maxval, const T* src,
+    T* dest, int64_t* indice_ptr, const int32_t padding_h, const int32_t padding_w,
+    const int64_t n_batch, const int64_t n_channel, const int64_t x_height, const int64_t x_width,
+    const int64_t y_height, const int64_t y_width, const int32_t kernel_size_h,
+    const int32_t kernel_size_w, const int32_t stride_h, const int32_t stride_w,
+    const int32_t dilation_h, const int32_t dilation_w, const bool return_indices,
+    const bool ceil_mode) {
```
Comment (leaves-zwx) [inline oneflow/user/kernels/pooling_kernel_util.h:144]: 这个 ceil_mode 没用到

File: oneflow/user/kernels/pooling_kernel_util.h [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+          max_value = val;
+          maxindex = tcntr;
+          src_idx = search_idx;
+        }
+      }
+    }
+    /* set output to local max */
+    dest[dest_idx] = src[src_idx];
+    /* store location of max */
+    indice_ptr[indp] = maxindex;
```
Comment (leaves-zwx) [inline oneflow/user/kernels/pooling_kernel_util.h:195]: 这里难道不是
```
dest[num] = src[src_idx];
indice_ptr[num] = maxindex;
```
就可以了吗？
前面求 xstart, ystart 是干啥？
如果要 src 的 plane 偏移不是直接
```
(n * n_channel + c) * x_height * y_width
```
就可以了？

File: oneflow/user/kernels/pooling_kernel_util.h [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+                                const int64_t* indice_ptr, const int64_t n_batch,
+                                const int64_t n_channel, const int64_t src_time,
+                                const int64_t src_height, const int64_t src_width,
+                                const int64_t dst_time, const int64_t dst_height,
+                                const int64_t dst_width, const bool return_indices,
+                                const bool ceil_mode);
+};
+
+template<typename T>
+OF_DEVICE_FUNC void Maxpool2dFarwardCompute(
```
Comment (leaves-zwx) [inline oneflow/user/kernels/pooling_kernel_util.h:137]: pytorch 里面对 nhwc 型的 tensor 做了特殊的优化，优化了写性能，使用 shared memory 来把写输出变成连续的内存访问，这里面没看到对等的实现

File: oneflow/user/kernels/pooling_kernel_util.h (hunk 4 above)
Comment (leaves-zwx) [reply oneflow/user/kernels/pooling_kernel_util.h:144]: 参数里面既然有了 index_helper，按照道理是不需要 y_height 和 y_width

Comment (leaves-zwx) [reply oneflow/user/kernels/pooling_kernel_util.h:144]: 如果参数太多可以包成结构体来减低复杂度

File: oneflow/user/kernels/pooling_kernel.cpp [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+    Memset<device_type>(ctx->device_ctx(), dest, 0, out_bytes_size);
+
+    PoolingKernelUtil<device_type, T>::Maxpool2dBackward(
+        ctx->device_ctx(), index_helper, elem_num, src, dest, indice_ptr, n_batch, n_channel,
+        src_height, src_width, dst_height, dst_width, return_indices, ceil_mode);
+  };
+};
+
+template<DeviceType device_type, typename T>
+class MaxPool3dKernel final : public user_op::OpKernel {
```
Comment (leaves-zwx) [inline oneflow/user/kernels/pooling_kernel.cpp:208]: 3d 与 2d 版本的 kernel 定义和 Util 函数里面有太多重复的代码，使用模版参数 `int NDIM` 消除一部分

File: oneflow/python/ops/nn_ops.py [hunk 8]
```diff
@@ -1522,6 +1582,423 @@ def calc_pool_padding(padding, dhw_offset, ndims):
...
+        padding = [padding, padding]
+    if len(padding) == 2:
+        if data_format == "NCHW":
+            padding = (0, 0, padding[0], padding[1])
+        elif data_format == "NHWC":
+            padding = (0, padding[0], padding[1], 0)
+        else:
+            raise ValueError('data_format must be "NHWC" or "NCHW".')
+
+    padding_type, pads_list = calc_pool_padding(padding, get_dhw_offset(channel_pos), 2)
```
Comment (leaves-zwx) [inline oneflow/python/ops/nn_ops.py:1829]: padding 的问题感觉可以通过外部来解决，不需要专门 "SAME", "VALID" 这种 padding_type，接口参数保持与 pytorch 相同就可以了
